engine/src/main/scala/cromwell/engine/workflow/WorkflowStoreActor.scala

```diff
@@ -48,70 +48,157 @@ object WorkflowStore {
   }
 
   /**
     * Retrieves up to n workflows which have not already been pulled into the engine and sets their pickedUp
     * flag to true
     */
-  def fetchRunnableWorkflows(n: Int): List[WorkflowToStart] = {
+  override def fetchRunnableWorkflows(n: Int)(implicit ec: ExecutionContext): Future[List[WorkflowToStart]] = {
     val startableWorkflows = workflowStore filter { _.state.isStartable } take n
     val updatedWorkflows = startableWorkflows map { _.copy(state = Running) }
     workflowStore = (workflowStore diff startableWorkflows) ++ updatedWorkflows
 
-    startableWorkflows map { _.toWorkflowToStart }
+    Future.successful(startableWorkflows map { _.toWorkflowToStart })
   }
 
-  def remove(id: WorkflowId): Boolean = {
+  override def remove(id: WorkflowId)(implicit ec: ExecutionContext): Future[Boolean] = {
     if (workflowStore.exists(_.id == id)) {
       workflowStore = workflowStore filterNot { _.id == id }
-      true
+      Future.successful(true)
     } else {
-      false
+      Future.successful(false)
     }
   }
+
+  override def initialize(implicit ec: ExecutionContext): Future[Unit] = Future.successful(())
 }
 
-class WorkflowStoreActor extends WorkflowStore with Actor with ActorLogging with ServiceRegistryClient {
-  /*
-    WARNING: WorkflowStore is NOT thread safe. Unless that statement is no longer true do NOT use threads
-    outside of the the single threaded Actor event loop
-   */
-
-  override def receive = {
-    case SubmitWorkflow(source) =>
-      val id = add(NonEmptyList(source)).head
-      registerIdWithMetadataService(id)
-      sender ! WorkflowSubmittedToStore(id)
-      log.info(s"Workflow $id submitted.")
-    case BatchSubmitWorkflows(sources) =>
-      val ids = add(sources)
-      ids foreach registerIdWithMetadataService
-      sender ! WorkflowsBatchSubmittedToStore(ids)
-      log.info(s"Workflows ${ids.list.mkString(", ")} submitted.")
-    case FetchRunnableWorkflows(n) => sender ! newWorkflowMessage(n)
-    case RemoveWorkflow(id) =>
-      if (!remove(id)) log.info(s"Attempted to remove ID $id from the WorkflowStore but it already exists!")
+case class WorkflowStoreActor(store: WorkflowStoreDatabase) extends LoggingFSM[WorkflowStoreActorState, WorkflowStoreActorData] with ActorLogging with ServiceRegistryClient {
+
+  implicit val ec: ExecutionContext = context.dispatcher
+
+  startWith(Idle, WorkflowStoreActorData(None, List.empty))
+  self ! InitializerCommand
+
+  when(Idle) {
+    case Event(cmd @ SubmitWorkflow(source), _) =>
+      val sndr = sender
+      val work = store.add(NonEmptyList(source)) map { ids =>
+        val id = ids.head
+        registerIdWithMetadataService(id)
+        sndr ! WorkflowSubmittedToStore(id)
+        log.info("Workflow {} submitted.", id)
+      }
+      addWorkCompletionHooks(cmd, work)
+      goto(Working) using stateData.copy(currentOperation = Option(cmd))
+    case Event(cmd @ BatchSubmitWorkflows(sources), _) =>
```

Comment:
this case and the preceding feel like they could be collapsed with a judiciously chosen common trait